File: LeetCode/0456. 132 Pattern/solution.py
"""
102 / 102 test cases passed.
Runtime: 68 ms
Memory Usage: 27.1 MB
"""
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        minn = float('-inf')
        stk = []
        for num in reversed(nums):
            if num < minn: return True
            while stk and stk[-1] < num:
                minn = stk.pop()
            stk.append(num)
        return False
        

File: LeetCode/0456. 132 Pattern/solution.cpp
/**
 * 102 / 102 test cases passed.
 * Runtime: 56 ms
 * Memory Usage: 37.2 MB 
 */
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int minn = INT_MIN;
        stack<int> stk;
        for (int i = nums.size() - 1; i >= 0; -- i) {
            if (nums[i] < minn) return true;
            while (!stk.empty() && stk.top() < nums[i]) {
                minn = stk.top();
                stk.pop();
            }
            stk.push(nums[i]);
        }
        return false;
    }
};
